File: structy/Graphs/longest_path.py
# Write a function, longest_path, that takes in an adjacency list for a directed acyclic graph. 
# The function should return the length of the longest path within the graph. A path may start and end at any
#  two nodes. The length of a path is considered the number of edges in the path, not the number of nodes.

graph = {
  'a': ['c', 'b'],
  'b': ['c'],
  'c': []
}

def longest_path(graph):
    distance = {}
    for node in graph:
        if len(graph[node]) == 0:
            distance[node] = 0

    for node in graph:
        traverse(graph, node, distance)
    
    return max(distance.values())


def traverse(grid, node, distance):
    if node in distance:
        return distance[node]

    longest = 0
    for neighbor in graph[node]:
        length = traverse(grid, neighbor, distance)
        if length > longest:
            longest = length 

    distance[node] = 1 + longest
    return distance[node]


print(longest_path(graph)) # -> 2

File: structy/Graphs/longestPath.js
function longestPath(graph) {
    const distance = {};
    for (const node in graph) {
        if (graph[node].length === 0) {
            distance[node] = 0;
        }
    };

    for (const node in graph) {
        explore(graph, node, distance);
    };

    return Math.max(...Object.values(distance))
};

function explore(graph, node, distance) {
    if (node in distance) return distance[node];

    let max = 0;
    for (const neighbor of graph[node]) {
        const size = explore(graph, neighbor, distance);
        if (size > max) max = size
    };

    distance[node] = 1 + max;
    return distance[node];
};

const graph = {
    a: ['c', 'b'],
    b: ['c'],
    c: []
  };
  
console.log(longestPath(graph)); // -> 2
